src/component.jl

```diff
@@ -302,3 +302,201 @@ function _shared_component(typedef, mod)
+        push!(c.indices, eid)
+        push!(c.group, length(c.group)+1)
+        push!(c.group_size, 1)
+        push!(c.data, v)
+    end
+    return v
+end
+
+# c[entity] = parent
+# set the value of this entity to that of parent
+@inline function Base.setindex!(c::GroupedComponent, p::Entity, e::Entity)
+    @boundscheck if !in(p, c)
+        throw(BoundsError(c, p))
+    end
+    pg = c.group[c.indices[p.id]]
+    if in(e, c)
+        eg = c.group[c.indices[e.id]]
+        if c.group_size[eg] == 1 #is_unique_in(c.group[c.indices[eid]], c.group)
+            # if this entity is the only one holding onto a value, remove that 
+            # value and cleanup group indices
+            deleteat!(c.data, idx)
+            deleteat!(c.group_size, idx)
+            for i in eachindex(c.group)
+                c.group[i] = c.group[i] - (c.group[i] > eg)
+            end
+        else
+            c.group_size[eg] -= 1
+        end
+        # adjust group index either way
+        c.group[c.indices[e.id]] = pg
+    else
+        # if the entity is not in there we have to add it
+        push!(c.indices, e.id)
+        push!(c.group, c.group[c.indices[p.id]])
+    end
+    c.group_size[pg] += 1
+
+    return c[p]
+end
+
+# c[ParentGroup(entity)] = value
+# set the value for all entities grouped with entity
+@inline function Base.setindex!(c::GroupedComponent{T}, v::T, x::ApplyToGroup) where {T}
+    e = x.e
+    @boundscheck if !in(e, c)
+        throw(BoundsError(c, e))
+    end
+    @inbounds c.data[c.group[c.indices[e.id]]] = v
+    return v
+end
+
+
+Base.length(c::GroupedComponent) = length(c.group)
+
+function Base.empty!(c::GroupedComponent)
+    empty!(c.indices)
+    empty!(c.group)
+    empty!(c.group_size)
+    empty!(c.data)
+    return c
+end
+
+
+function Base.pop!(c::GroupedComponent, e::Entity)
+    @boundscheck if !in(e, c)
+        throw(BoundsError(c, e))
+    end
+
+    @inbounds begin
+        id = c.indices[e.id]
+        g = c.group[id]
+
+        c.group[id] = c.group[end]
+        c.group_size[g] -= 1
+        pop!(c.group)
+        pop!(c.indices, e.id)
+
+        val = c.data[g]
+
+        if c.group_size[g] == 0
+            deleteat!(c.data, g)
+            deleteat!(c.group_size, g)
+            for i in eachindex(c.group)
+                if c.group[i] > g
+                    c.group[i] -= 1
+                end
+            end
+        end
+    end
+
+    return val
+end
+
+@inline Base.iterate(c::GroupedComponent, args...) = iterate(c.data, args...)
+Base.sortperm(c::GroupedComponent) = sortperm(c.group)
+
+
+macro grouped_component(typedef)
+    return esc(Overseer._grouped_component(typedef, __module__))
+end
+
+function _grouped_component(typedef, mod)
+    t = process_typedef(typedef, mod)
+    t1, tn = t 
+    return quote
+        $t1
+       	Overseer.component_type(::Type{$tn}) = Overseer.GroupedComponent
+    end
+end
+
+function make_unique!(c::GroupedComponent)
```

Comment:
probably here some @inbounds can also be sprinkled around for some performance gains